components/core/src/clp_s/ArchiveWriter.cpp

```diff
@@ -34,48 +37,36 @@ void ArchiveWriter::open(ArchiveWriterOption const& option) {
     m_array_dict = std::make_shared<LogTypeDictionaryWriter>();
     m_array_dict->open(array_dict_path, m_compression_level, UINT64_MAX);
 
-    std::string timestamp_local_dict_path = m_archive_path + constants::cArchiveTimestampDictFile;
-    m_timestamp_dict->open_local(timestamp_local_dict_path, m_compression_level);
+    std::string timestamp_dict_path = m_archive_path + constants::cArchiveTimestampDictFile;
+    m_timestamp_dict = std::make_shared<TimestampDictionaryWriter>();
+    m_timestamp_dict->open(timestamp_dict_path, m_compression_level);
 }
 
-size_t ArchiveWriter::close() {
-    size_t compressed_size{0};
-    compressed_size += m_var_dict->close();
-    compressed_size += m_log_dict->close();
-    compressed_size += m_array_dict->close();
-    compressed_size += m_timestamp_dict->close_local();
-
-    m_tables_file_writer.open(
-            m_archive_path + constants::cArchiveTablesFile,
-            FileWriter::OpenMode::CreateForWriting
-    );
-    m_table_metadata_file_writer.open(
-            m_archive_path + constants::cArchiveTableMetadataFile,
-            FileWriter::OpenMode::CreateForWriting
-    );
-    m_table_metadata_compressor.open(m_table_metadata_file_writer, m_compression_level);
-    m_table_metadata_compressor.write_numeric_value(m_id_to_schema_writer.size());
-    for (auto& i : m_id_to_schema_writer) {
-        m_table_metadata_compressor.write_numeric_value(i.first);
-        m_table_metadata_compressor.write_numeric_value(i.second->get_num_messages());
-        m_table_metadata_compressor.write_numeric_value(m_tables_file_writer.get_pos());
-
-        m_tables_compressor.open(m_tables_file_writer, m_compression_level);
-        i.second->store(m_tables_compressor);
-        m_tables_compressor.close();
-        delete i.second;
+void ArchiveWriter::close() {
+    m_compressed_size += m_var_dict->close();
+    m_compressed_size += m_log_dict->close();
+    m_compressed_size += m_array_dict->close();
+    m_compressed_size += m_timestamp_dict->close();
+    m_compressed_size += m_schema_tree.store(m_archive_path, m_compression_level);
+    m_compressed_size += m_schema_map.store(m_archive_path, m_compression_level);
+    m_compressed_size += store_tables();
+
+    if (m_metadata_db) {
+        m_metadata_db->open();
+        update_metadata_db();
+        m_metadata_db->close();
     }
```

Comment:
I think we could move the if-condition and the open-close into `update_metadata_db` to simplify the code; otherwise `update_metadata_db` depends on `m_metadata_db` being valid and open which sounds a bit error prone.